source/common/http/http1/codec_impl.cc

```diff
@@ -247,6 +247,15 @@ void RequestStreamEncoderImpl::encodeHeaders(const HeaderMap& headers, bool end_
 
   if (method->value() == Headers::get().MethodValues.Head.c_str()) {
     head_request_ = true;
+    // If this is an incomplete request (end_stream=false), then connection_.onEncodeComplete();
+    // will not be called in StreamEncoderImpl::encodeHeaders, head request state will not be passed
+    // to the corresponding response so here need to call this method
+    if (!end_stream) {
+      auto c = dynamic_cast<ClientConnectionImpl*>(&connection_);
+      if (c) {
```

Comment:
This should always be true. If we are going to do the dynamic cast we can just assume it is valid, though I'm not a huge fan of dynamic_cast for something like this. The other option is to store `ClientConncetion` in `RequestStreamEncoderImpl` (and server connection in `ResponseStreamEncoderImpl` and then have a base class virtual function `connection()` which returns the connection base class. @alyssawilk any opinions here?

Reply:
I'd suggest that just like we have a virtual onEncodeComplete that the client/server connections implement, we could have a pure virtual onEncodeHeaders.  We could handle passing on head status in the ClientConnectionImpl and the ServerConnectionImpl could just be a no-op for now.

I think that would be consistent with the style of the rest of the code base, and make it clear what part of the request/response pipeline is being handled.